src/cargo/core/resolver/conflict_cache.rs

```diff
@@ -1,40 +1,98 @@
-use std::collections::{HashMap, HashSet};
+use std::collections::{BTreeMap, HashMap, HashSet};
 
 use super::types::ConflictReason;
 use core::resolver::Context;
 use core::{Dependency, PackageId};
 
+enum ConflictStore {
+    Con(BTreeMap<PackageId, ConflictReason>),
+    Map(HashMap<PackageId, ConflictStore>),
+}
+
+impl ConflictStore {
+    /// Finds any known set of conflicts, if any,
+    /// which are activated in `cx` and pass the `filter` specified?
+    pub fn find_conflicting<F>(
+        &self,
+        cx: &Context,
+        filter: &F,
+    ) -> Option<&BTreeMap<PackageId, ConflictReason>>
+    where
+        for<'r> F: Fn(&'r &BTreeMap<PackageId, ConflictReason>) -> bool,
+    {
+        match self {
+            ConflictStore::Con(c) => {
+                if filter(&c) {
+                    Some(c)
+                } else {
+                    None
+                }
+            }
+            ConflictStore::Map(m) => {
+                for (pid, store) in m {
+                    if cx.is_active(pid) {
+                        if let Some(o) = store.find_conflicting(cx, filter) {
+                            return Some(o);
+                        }
+                    }
+                }
+                None
+            }
+        }
+    }
+
+    pub fn insert<'a>(
+        &mut self,
+        mut iter: impl Iterator<Item = &'a PackageId>,
+        con: BTreeMap<PackageId, ConflictReason>,
+    ) {
+        if let Some(pid) = iter.next() {
+            if let ConflictStore::Map(p) = self {
+                p.entry(pid.clone())
+                    .or_insert_with(|| ConflictStore::Map(HashMap::new()))
+                    .insert(iter, con);
+            }
+            // else, We already have a subset of this in the ConflictStore
+        } else {
+            *self = ConflictStore::Con(con)
```

Comment:
This whole method threw me for a spin when I first read it, but I think I'm starting to understand it now. Basically this data structure is sort of an optimized "set of sets", but not quite. The path to a leaf (currently `Con`) will be the ordered traversal of the leaf node's set, so `Con` is as deep in the tree as it has elements in its set.

The real tricky part here is that let's say we add a leaf where the set of keys is a subset of keys in an existing node. This means we found a conflict with *less* packages than before (which seems odd, but hey I guess it happens). It's therefore safe to prune the entire subtree, regardless of how many leaves it has, because we found a smaller and more applicable set anyway.

Does that sound right? (if so, could that be codified in a comment?)